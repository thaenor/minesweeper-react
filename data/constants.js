export const total_mines = 20;
export const size = 10;
