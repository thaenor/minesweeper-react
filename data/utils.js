import { size } from './constants';

export function is_left_edge(pos) {
  return pos % size === 0;
}

export function is_right_edge(pos) {
  return pos % size === size - 1;
}

export function are_all_tiles_revealed(board) {
  let filled_tiles = board.filter(tile => tile.revealed == true || tile.type === 'bomb');

  return (filled_tiles.length === size*size);
}

export function reveal_all_tiles(board) {
  let all_tiles = [...board];
  return all_tiles.map( til => {
    til.revealed = true;
    return til;
  });
}
