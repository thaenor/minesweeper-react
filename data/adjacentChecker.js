import { size } from '../data/constants';
import { is_left_edge, is_right_edge } from '../data/utils';

async function check_tile(tile, board) {
  if (tile.type === 'empty' && tile.warning === 0 && tile.revealed === false) {
    reveal_adjacent_tiles(board, tile);
  }
  tile.revealed = true;
}

export async function reveal_adjacent_tiles(board, tile) {
  const i = tile.pos;
  const b = board;
  const s = size;

  //West
  if (i > 0 && !is_left_edge(i)) {
    await check_tile(b[i - 1], b);
  }

  //NE
  if (i > s - 1 && !is_right_edge(i)) {
    await check_tile(b[i + 1 - s], b);
  }

  //North
  if (i > s) {
    await check_tile(b[i - s], b);
  }

  //NW
  if (i > s + 1 && !is_left_edge(i)) {
    await check_tile(b[i - 1 - s], b);
  }

  //East
  if (i < s * s - 2 && !is_right_edge(i)) {
    await check_tile(b[i + 1], b);
  }

  //SW
  if (i < s * s - s && !is_left_edge(i)) {
    await check_tile(b[i - i + s], b);
  }

  //SE
  if (i < s * s - s - 2 && !is_right_edge(i)) {
    await check_tile(b[i + 1 + s], b);
  }

  //South
  if (i < s * s - s - 1) {
    await check_tile(b[i + s], b);
  }
}
