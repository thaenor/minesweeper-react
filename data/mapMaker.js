import { size, total_mines } from './constants';
import { is_left_edge, is_right_edge } from './utils';

export default function create_board() {
  let bomb_slots = Array(total_mines).fill({ type: 'bomb' });
  let empty_slots = Array(size * size - total_mines).fill({
    type: 'empty',
    warning: 0
  });
  let combined = empty_slots.concat(bomb_slots);
  let map = combined.sort(() => Math.random() - 0.5);
  const s = size;

  return map.map((tile, i, m) => {
    if (tile.type === 'empty') {
      let new_tile = { type: 'empty', warning: 0 };
      if (i > 0 && !is_left_edge(i) && m[i - 1].type === 'bomb')
        new_tile.warning++;
      if (i > s - 1 && !is_right_edge(i) && m[i + 1 - s].type === 'bomb')
        new_tile.warning++;
      if (i > s && m[i - s].type === 'bomb') new_tile.warning++;
      if (i > s + 1 && !is_left_edge(i) && m[i - 1 - s].type === 'bomb')
        new_tile.warning++;
      if (i < s * s - 2 && !is_right_edge(i) && m[i + 1].type === 'bomb')
        new_tile.warning++;
      if (i < s * s - s && !is_left_edge(i) && m[i - 1 + s].type === 'bomb')
        new_tile.warning++;
      if (
        i < s * s - s - 2 &&
        !is_right_edge(i) &&
        m[i + 1 + s].type === 'bomb'
      )
        new_tile.warning++;
      if (i < s * s - s - 1 && m[i + s].type === 'bomb') new_tile.warning++;

      new_tile.warning = new_tile.warning >= 4 ? 4 : new_tile.warning;
      return { ...new_tile, flag: false, revealed: false, pos: i };
    } else {
      return { ...tile, flag: false, revealed: false, pos: i };
    }
  });
}
