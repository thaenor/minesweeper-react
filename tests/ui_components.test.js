import React from 'react';
import renderer from 'react-test-renderer';

import { size, total_mines } from '../data/constants';
import Layout from '../components/layout';
import Desk from '../components/desk';
import Square from '../components/square';
import Mine from '../components/mine';
import Flag from '../components/flag';

describe('Map rendering', () => {
  test('Expect map to render with correct elements', () => {
    const component = renderer.create(
      <Layout title={`Minesweeper (active)`}>
        <Desk boardSize={size}>
          {[...Array(100).keys()].map(i => (
            <Square
              key={i}
              disabled={i === 55 || i === 10}
              onClick={e => select_square(e)}
            >
              {i === 10 && <Mine />}
              {i === 25 && <Flag />}
              {i === 77 ? '4' : ''}
            </Square>
          ))}
        </Desk>
      </Layout>
    );
    let tree = component.toJSON();
    const testInstance = component.root;

    expect(testInstance.findByType(Layout).props.title).toBe('Minesweeper (active)');
    expect(testInstance.findByType(Desk).props.boardSize).toBe(size);
    expect(tree).toMatchSnapshot();
  });
});
