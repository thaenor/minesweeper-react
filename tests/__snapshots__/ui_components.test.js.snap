// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map rendering Expect map to render with correct elements 1`] = `
<div
  className="a"
>
  <h1>
    Minesweeper (active)
  </h1>
  <div
    className="b c d e f"
  >
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="o b p i j k l m n"
      onClick={[Function]}
    >
      <svg
        viewBox="0 0 100 100"
        x="0px"
        y="0px"
      >
        <path
          d="M66.03,7.683c1.262-0.539,2.879-0.885,4.617-0.812c0.759,4.807-1.603,8.536-4.659,10.089  c-1.752,0.891-3.946,1.383-6.241,0.94c-2.073-0.4-3.62-1.285-4.918-2.607c-1.325-1.353-2.034-2.848-2.522-5.002  C51.867,8.348,50.5,6.976,48.46,6.614c-2.38-0.422-4.1,0.847-4.917,2.437c-0.939,1.828-0.262,3.975,0.128,6.2  c1.409-0.116,3.024-0.396,4.617-0.471c0.916,4.884,1.71,9.89,2.607,14.792c10.156,0.618,17.22,4.184,22.744,9.49  c5.384,5.173,9.56,12.2,10.432,21.76c0.967,10.606-2.938,18.79-7.481,24.453c-4.673,5.823-11.247,10.455-20.094,12.1  c-4.912,0.913-10.049,0.747-14.45-0.301c-4.317-1.026-8.073-2.798-11.201-4.915c-6.418-4.347-11.185-10.449-13.637-18.555  c-1.281-4.234-1.797-9.507-1.154-14.279c0.624-4.633,2.096-8.712,3.976-12.098c3.837-6.914,9.552-11.927,17.058-15.134  c-0.798-5.016-1.806-9.821-2.564-14.877c1.49-0.405,2.992-0.798,4.574-1.112c-0.288-1.806-0.763-3.39-0.727-5.173  c0.034-1.707,0.571-3.317,1.325-4.574c1.451-2.418,4.159-4.276,7.481-4.489c2.74-0.175,5.012,0.893,6.583,2.223  C55.424,5.5,56.4,7.45,56.925,9.735c0.53,2.313,2.5,3.856,4.873,3.676C64.427,13.212,66.671,10.963,66.03,7.683z"
          fill="#000000"
        />
      </svg>
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      <svg
        viewBox="0 0 48 60"
        x="0px"
        y="0px"
      >
        <g>
          <path
            d="M8.3,35.5V2.2c-0.1-1-0.9-1.7-1.9-1.7c-1.1,0-2,0.9-2,2v2.8v22v18.3c0,1.1,0.9,2,2,2s2-0.9,2-2L8.3,35.5L8.3,35.5z"
          />
          <path
            d="M43.3,26.3c-1.6-3.6-3.2-7.2-4.9-10.9c1.7-3.4,3.3-6.4,4.9-9.1c0.3-0.5,0.3-1.1,0.1-1.7s-0.7-1-1.3-1.2   c-2.5-0.7-5.1-1.1-8-1.1c-3.5,0-7.1,0.5-10.5,1c-3.4,0.5-6.6,1-9.9,1c-0.5,0-1.1,0-1.6,0v25.9c0.5,0,1.1,0,1.6,0   c3.5,0,7.1-0.5,10.5-1c3.4-0.5,6.6-1,9.9-1c2.5,0,4.8,0.3,6.9,0.9c0.2,0.1,0.4,0.1,0.6,0.1c0,0,0,0,0,0c1.1,0,2-0.9,2-2   C43.6,26.9,43.5,26.6,43.3,26.3z"
          />
        </g>
      </svg>
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="o b p i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      4
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
    <div
      className="g b h i j k l m n"
      onClick={[Function]}
    >
      
    </div>
  </div>
</div>
`;
